Refit breaks with interface inheritance where base doesn't specify refit methods

Forking from: https://github.com/reactiveui/refit/issues/650

Generic Interface:

```csharp
public interface IBase<T>
{
        public Task DoSomething(T foo);
}
```
Refit Interface:

```csharp
public interface IService  :  IBase<int>
{
        [Post("/api")]
        public Task DoSomething(int foo);
}
```

and got the same error message:

>   /InterfaceStubGenerator.Core/Refit.Generator.InterfaceStubGenerator/IService_refit.cs(16,11): error CS0535: 'Generated.FrobIService' does not implement interface member 'IBase<int>.DoSomething(int)' 

The issue is that the derived interface contains the refit attribute and we implement all expicitly. So the generated type doesn't actually implement the base.

--- a/InterfaceStubGenerator.Core/InterfaceStubGenerator.cs
+++ b/InterfaceStubGenerator.Core/InterfaceStubGenerator.cs
@@ -259,13 +259,19 @@ namespace Refit.Implementation
             var derivedNonRefitMethods = derivedMethods.Except(derivedMethods, SymbolEqualityComparer.Default).Cast<IMethodSymbol>().ToList();
 
             // Handle Refit Methods            
-            foreach(var method in refitMethods.Concat(derivedRefitMethods))
+            foreach(var method in refitMethods)
             {
-                ProcessRefitMethod(source, method);
+                ProcessRefitMethod(source, method, true);
             }
 
+            foreach (var method in refitMethods.Concat(derivedRefitMethods))
+            {
+                ProcessRefitMethod(source, method, false);
+            }
+
+
             // Handle non-refit Methods that aren't static or properties or have a method body
-            foreach(var method in nonRefitMethods.Concat(derivedNonRefitMethods))
+            foreach (var method in nonRefitMethods.Concat(derivedNonRefitMethods))
             {
                 if (method.IsStatic ||
                     method.MethodKind == MethodKind.PropertyGet ||
@@ -292,9 +298,15 @@ namespace Refit.Implementation
             return source.ToString();
         }
 
-        void ProcessRefitMethod(StringBuilder source, IMethodSymbol methodSymbol)
+        /// <summary>
+        /// Generates the body of the Refit method
+        /// </summary>
+        /// <param name="source"></param>
+        /// <param name="methodSymbol"></param>
+        /// <param name="isTopLevel">True if directly from the type we're generating for, false for methods found on base interfaces</param>
+        void ProcessRefitMethod(StringBuilder source, IMethodSymbol methodSymbol, bool isTopLevel)
         {
-            WriteMethodOpening(source, methodSymbol);
+            WriteMethodOpening(source, methodSymbol, !isTopLevel);
 
             // Build the list of args for the array
             var argList = new List<string>();
@@ -330,7 +342,7 @@ namespace Refit.Implementation
 
         void ProcessDisposableMethod(StringBuilder source, IMethodSymbol methodSymbol)
         {
-            WriteMethodOpening(source, methodSymbol);
+            WriteMethodOpening(source, methodSymbol, true);
 
             source.Append(@"
                 Client?.Dispose();
@@ -396,7 +408,7 @@ namespace Refit.Implementation
 
         void ProcessNonRefitMethod(StringBuilder source, IMethodSymbol methodSymbol, GeneratorExecutionContext context)
         {
-            WriteMethodOpening(source, methodSymbol);
+            WriteMethodOpening(source, methodSymbol, true);
 
             source.Append(@"
                 throw new global::System.NotImplementedException(""Either this method has no Refit HTTP method attribute or you've used something other than a string literal for the 'path' argument."");
@@ -411,12 +423,20 @@ namespace Refit.Implementation
             }            
         }
 
-        void WriteMethodOpening(StringBuilder source, IMethodSymbol methodSymbol)
+        void WriteMethodOpening(StringBuilder source, IMethodSymbol methodSymbol, bool isExplicitInterface)
         {
+            var visibility = !isExplicitInterface ? "public " : string.Empty;
+
             source.Append(@$"
 
         /// <inheritdoc />
-        {methodSymbol.ReturnType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)} {methodSymbol.ContainingType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}.{methodSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}(");
+        {visibility}{methodSymbol.ReturnType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)} ");
+
+            if(isExplicitInterface)
+            {
+                source.Append(@$"{methodSymbol.ContainingType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}.");
+            }
+            source.Append(@$"{methodSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}(");            
 
             if(methodSymbol.Parameters.Length > 0)
             {

--- a/Refit.Tests/InheritedInterfacesApi.cs
+++ b/Refit.Tests/InheritedInterfacesApi.cs
@@ -52,6 +52,24 @@ namespace Refit.Tests
     {
 
     }
+
+    public interface IAmInterfaceEWithNoRefit<T>
+    {
+        public Task DoSomething(T parameter);
+
+        public Task DoSomethingElse();
+    }
+
+    public interface IImplementTheInterfaceAndUseRefit : IAmInterfaceEWithNoRefit<int>
+    {
+#pragma warning disable CS0108 // Member hides inherited member; missing new keyword
+        [Get("/doSomething")]
+        public Task DoSomething(int parameter);
+#pragma warning restore CS0108 // Member hides inherited member; missing new keyword
+
+        [Get("/DoSomethingElse")]
+        public new Task DoSomethingElse();
+    }
 }
 
 namespace Refit.Tests.SeparateNamespaceWithModel

--- a/Refit.Tests/InterfaceStubGenerator.cs
+++ b/Refit.Tests/InterfaceStubGenerator.cs
@@ -175,6 +175,110 @@ namespace Refit.Implementation
 
 
         /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.User> GetUser(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUser"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<global::Refit.Tests.User> GetUserObservable(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserObservable"", new global::System.Type[] { typeof(string) } );
+            return (global::System.IObservable<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<global::Refit.Tests.User> GetUserCamelCase(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserCamelCase"", new global::System.Type[] { typeof(string) } );
+            return (global::System.IObservable<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::System.Collections.Generic.List<global::Refit.Tests.User>> GetOrgMembers(string @orgName, global::System.Threading.CancellationToken @cancellationToken) 
+        {
+            var ______arguments = new object[] { @orgName, @cancellationToken };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetOrgMembers"", new global::System.Type[] { typeof(string), typeof(global::System.Threading.CancellationToken) } );
+            return (global::System.Threading.Tasks.Task<global::System.Collections.Generic.List<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.UserSearchResult> FindUsers(string @q) 
+        {
+            var ______arguments = new object[] { @q };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""FindUsers"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.UserSearchResult>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::System.Net.Http.HttpResponseMessage> GetIndex() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetIndex"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task<global::System.Net.Http.HttpResponseMessage>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<string> GetIndexObservable() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetIndexObservable"", new global::System.Type[] {  } );
+            return (global::System.IObservable<string>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.User> NothingToSeeHere() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""NothingToSeeHere"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>> NothingToSeeHereWithMetadata() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""NothingToSeeHereWithMetadata"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>> GetUserWithMetadata(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserWithMetadata"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<global::Refit.ApiResponse<global::Refit.Tests.User>> GetUserObservableWithMetadata(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserObservableWithMetadata"", new global::System.Type[] { typeof(string) } );
+            return (global::System.IObservable<global::Refit.ApiResponse<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.User> CreateUser(global::Refit.Tests.User @user) 
+        {
+            var ______arguments = new object[] { @user };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""CreateUser"", new global::System.Type[] { typeof(global::Refit.Tests.User) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>> CreateUserWithMetadata(global::Refit.Tests.User @user) 
+        {
+            var ______arguments = new object[] { @user };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""CreateUserWithMetadata"", new global::System.Type[] { typeof(global::Refit.Tests.User) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.ApiResponse<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
         global::System.Threading.Tasks.Task<global::Refit.Tests.User> global::Refit.Tests.IGitHubApi.GetUser(string @userName) 
         {
             var ______arguments = new object[] { @userName };
@@ -315,6 +419,14 @@ namespace Refit.Implementation
 
 
         /// <inheritdoc />
+        public global::System.Threading.Tasks.Task RefitMethod() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""RefitMethod"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
         global::System.Threading.Tasks.Task global::Refit.Tests.IGitHubApiDisposable.RefitMethod() 
         {
             var ______arguments = new object[] {  };
@@ -363,6 +475,70 @@ namespace Refit.Implementation
         }
     
 
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.User> GetUser(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUser"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<global::Refit.Tests.User> GetUserObservable(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserObservable"", new global::System.Type[] { typeof(string) } );
+            return (global::System.IObservable<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<global::Refit.Tests.User> GetUserCamelCase(string @userName) 
+        {
+            var ______arguments = new object[] { @userName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetUserCamelCase"", new global::System.Type[] { typeof(string) } );
+            return (global::System.IObservable<global::Refit.Tests.User>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::System.Collections.Generic.List<global::Refit.Tests.User>> GetOrgMembers(string @orgName) 
+        {
+            var ______arguments = new object[] { @orgName };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetOrgMembers"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::System.Collections.Generic.List<global::Refit.Tests.User>>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::Refit.Tests.UserSearchResult> FindUsers(string @q) 
+        {
+            var ______arguments = new object[] { @q };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""FindUsers"", new global::System.Type[] { typeof(string) } );
+            return (global::System.Threading.Tasks.Task<global::Refit.Tests.UserSearchResult>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task<global::System.Net.Http.HttpResponseMessage> GetIndex() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetIndex"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task<global::System.Net.Http.HttpResponseMessage>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.IObservable<string> GetIndexObservable() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetIndexObservable"", new global::System.Type[] {  } );
+            return (global::System.IObservable<string>)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task NothingToSeeHere() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""NothingToSeeHere"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task)______func(this.Client, ______arguments);
+        }
 
         /// <inheritdoc />
         global::System.Threading.Tasks.Task<global::Refit.Tests.User> global::Refit.Tests.TestNested.INestedGitHubApi.GetUser(string @userName) 
@@ -529,6 +705,22 @@ namespace Refit.Implementation
         }
     
 
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task GetRoot() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""GetRoot"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task)______func(this.Client, ______arguments);
+        }
+
+        /// <inheritdoc />
+        public global::System.Threading.Tasks.Task PostRoot() 
+        {
+            var ______arguments = new object[] {  };
+            var ______func = requestBuilder.BuildRestResultFuncForMethod(""PostRoot"", new global::System.Type[] {  } );
+            return (global::System.Threading.Tasks.Task)______func(this.Client, ______arguments);
+        }
 
         /// <inheritdoc />
         global::System.Threading.Tasks.Task global::IServiceWithoutNamespace.GetRoot() 

--- a/Refit.Tests/RestService.cs
+++ b/Refit.Tests/RestService.cs
@@ -1802,6 +1802,44 @@ namespace Refit.Tests
             mockHttp.VerifyNoOutstandingExpectation();
         }
 
+
+        [Fact]
+        public async Task InheritedInterfaceWithoutRefitInBaseMethodsTest()
+        {
+            var mockHttp = new MockHttpMessageHandler();
+
+            var settings = new RefitSettings
+            {
+                HttpMessageHandlerFactory = () => mockHttp
+            };
+
+            var fixture = RestService.For<IImplementTheInterfaceAndUseRefit>("https://httpbin.org", settings);
+
+            mockHttp.Expect(HttpMethod.Get, "https://httpbin.org/doSomething")
+                    .WithQueryString("parameter", "4")
+                    .Respond("application/json", nameof(IImplementTheInterfaceAndUseRefit.DoSomething));
+
+            await fixture.DoSomething(4);
+            mockHttp.VerifyNoOutstandingExpectation();
+
+
+
+
+            mockHttp.Expect(HttpMethod.Get, "https://httpbin.org/DoSomethingElse")
+                .Respond("application/json", nameof(IImplementTheInterfaceAndUseRefit.DoSomethingElse));
+            await fixture.DoSomethingElse();
+            mockHttp.VerifyNoOutstandingExpectation();
+
+
+            mockHttp.Expect(HttpMethod.Get, "https://httpbin.org/DoSomethingElse")
+    .Respond("application/json", nameof(IImplementTheInterfaceAndUseRefit.DoSomethingElse));
+            await ((IAmInterfaceEWithNoRefit<int>)fixture).DoSomethingElse();
+            mockHttp.VerifyNoOutstandingExpectation();
+
+
+            Assert.Throws<InvalidOperationException>(() => RestService.For<IAmInterfaceEWithNoRefit<int>>("https://httpbin.org"));
+        }
+
         [Fact]
         public async Task DictionaryDynamicQueryparametersTest()
         {

